File: Tests/Test_Hardlinks.vb
Option Explicit On
Option Strict On
Option Compare Binary
Option Infer Off

Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq

Namespace Tests
    Module Tests_Hardlinks
        Function Test_Hardlinks1(rootTestFolder As String) As Boolean
            Using testFileSource As New DisposableFile(Path.Combine(rootTestFolder, "hardlinks1Source.txt"))
                File.WriteAllText(testFileSource, "testText")
                Dim hardlinkPath As String = Path.Combine(rootTestFolder, "hardlinks1.txt")
                Environment.CurrentDirectory = rootTestFolder

                WalkmanLib.CreateHardLink(hardlinkPath, "hardlinks1Source.txt")

                Using testHardlink As New DisposableFile(hardlinkPath, False)
                    Return TestString("Hardlinks1", File.ReadAllText(hardlinkPath), "testText")
                End Using
            End Using
        End Function

        Function Test_Hardlinks2(rootTestFolder As String) As Boolean
            Using testFileSource As New DisposableFile(Path.Combine(rootTestFolder, "hardlinks2Source.txt"))
                File.WriteAllText(testFileSource, "testText")
                Dim hardlinkPath As String = Path.Combine(rootTestFolder, "hardlinks2.txt")

                WalkmanLib.CreateHardLink(hardlinkPath, testFileSource)

                Using testHardlink As New DisposableFile(hardlinkPath, False)
                    Return TestString("Hardlinks2", File.ReadAllText(hardlinkPath), "testText")
                End Using
            End Using
        End Function

        Function Test_Hardlinks3(rootTestFolder As String) As Boolean
            Using testDirRoot As New DisposableDirectory(Path.Combine(rootTestFolder, "hardlinks3Root")),
                  testFileSource As New DisposableFile(Path.Combine(testDirRoot, "hardlinks3Source.txt"))

                File.WriteAllText(testFileSource, "testText")
                Dim hardlinkPath As String = Path.Combine(rootTestFolder, "hardlinks3.txt")
                Dim hardlinkTarget As String = Path.Combine(Path.GetFileName(testDirRoot), Path.GetFileName(testFileSource))
                Environment.CurrentDirectory = rootTestFolder

                WalkmanLib.CreateHardLink(hardlinkPath, hardlinkTarget)

                Using testHardlink As New DisposableFile(hardlinkPath, False)
                    Return TestString("Hardlinks3", File.ReadAllText(hardlinkPath), "testText")
                End Using
            End Using
        End Function

        Function Test_Hardlinks4(rootTestFolder As String) As Boolean
            Using testFileSource As New DisposableFile(Path.Combine(rootTestFolder, "hardlinks4Source.txt")),
                  testDirTarget As New DisposableDirectory(Path.Combine(rootTestFolder, "hardlinks4Target"))

                File.WriteAllText(testFileSource, "testText")
                Dim hardlinkPath As String = Path.Combine(testDirTarget, "hardlinks4.txt")
                Dim hardlinkTarget As String = Path.Combine("..", Path.GetFileName(testFileSource))
                Environment.CurrentDirectory = testDirTarget

                WalkmanLib.CreateHardLink(hardlinkPath, hardlinkTarget)

                Environment.CurrentDirectory = rootTestFolder ' allow testDirTarget to be deleted
                Using testHardlink As New DisposableFile(hardlinkPath, False)
                    Return TestString("Hardlinks4", File.ReadAllText(hardlinkPath), "testText")
                End Using
            End Using
        End Function

        Function Test_Hardlinks5(rootTestFolder As String) As Boolean
            Using testDirSource As New DisposableDirectory(Path.Combine(rootTestFolder, "hardlinks5Source")),
                  testFileSource As New DisposableFile(Path.Combine(testDirSource, "hardlinks5Source.txt")),
                  testDirTarget As New DisposableDirectory(Path.Combine(rootTestFolder, "hardlinks5Target"))

                File.WriteAllText(testFileSource, "testText")
                Dim hardlinkPath As String = Path.Combine(testDirTarget, "hardlinks5.txt")
                Dim hardlinkTarget As String = Path.Combine("..", Path.GetFileName(testDirSource), Path.GetFileName(testFileSource))
                Environment.CurrentDirectory = testDirTarget

                WalkmanLib.CreateHardLink(hardlinkPath, hardlinkTarget)

                Environment.CurrentDirectory = rootTestFolder ' allow testDirTarget to be deleted
                Using testHardlink As New DisposableFile(hardlinkPath, False)
                    Return TestString("Hardlinks5", File.ReadAllText(hardlinkPath), "testText")
                End Using
            End Using
        End Function

        Function Test_HardlinkThrows1(rootTestFolder As String) As Boolean
            Dim ex As Exception = New NoException
            Try
                WalkmanLib.CreateHardLink(Path.Combine(rootTestFolder, "hardlinkThrows1.txt"), "nonExistantFile.txt")
            Catch ex2 As Exception
                ex = ex2
            End Try
            Return TestType("HardlinkThrows1", ex.GetType, GetType(FileNotFoundException))
        End Function

        Function Test_HardlinkThrows2(rootTestFolder As String) As Boolean
            Using testFile As New DisposableFile(Path.Combine(rootTestFolder, "hardlinkThrows2Source.txt"))
                Dim ex As Exception = New NoException
                Try
                    WalkmanLib.CreateHardLink(Path.Combine(rootTestFolder, "nonExistantFolder", "hardlinkThrows2.txt"), testFile)
                Catch ex2 As Exception
                    ex = ex2
                End Try
                Return TestType("HardlinkThrows2", ex.GetType, GetType(DirectoryNotFoundException))
            End Using
        End Function

        Function Test_HardlinkThrows3(rootTestFolder As String) As Boolean
            Using testSource As New DisposableFile(Path.Combine(rootTestFolder, "hardlinkThrows3Source.txt")),
                  testFile As New DisposableFile(Path.Combine(rootTestFolder, "hardlinkThrows3.txt"))
                Dim ex As Exception = New NoException
                Try
                    WalkmanLib.CreateHardLink(testFile, testSource)
                Catch ex2 As Exception
                    ex = ex2
                End Try
                Return TestType("HardlinkThrows3", ex.GetType, GetType(IOException))
            End Using
        End Function

        Function Test_HardlinkThrows4(rootTestFolder As String) As Boolean
            Using testFile As New DisposableFile(Path.Combine(rootTestFolder, "hardlinkThrows4Source.txt"))
                Dim ex As Exception = New NoException
                Try
                    WalkmanLib.CreateHardLink(Path.Combine(Environment.SystemDirectory, "symlinkThrows3.txt"), testFile)
                Catch ex2 As Exception
                    ex = ex2
                End Try
                Return TestType("HardlinkThrows4", ex.GetType, GetType(UnauthorizedAccessException))
            End Using
        End Function

        Function Test_Hardlink6(rootTestFolder As String) As Boolean
            Using testFileSource As New DisposableFile(Path.Combine(rootTestFolder, "hardlinks6Source.txt"))
                Dim hardlinkPath As String = Path.Combine(rootTestFolder, "hardlinks6.txt")

                WalkmanLib.CreateHardLink(hardlinkPath, testFileSource)

                Using testHardlink As New DisposableFile(hardlinkPath, False)
                    Return TestNumber("Hardlinks6", WalkmanLib.GetHardlinkCount(hardlinkPath), 2)
                End Using
            End Using
        End Function

        Function Test_Hardlink7(rootTestFolder As String) As Boolean
            Using testFileSource As New DisposableFile(Path.Combine(rootTestFolder, "hardlinks7Source.txt"))
                Dim hardlinkPath As String = Path.Combine(rootTestFolder, "hardlinks7.txt")

                WalkmanLib.CreateHardLink(hardlinkPath, testFileSource)

                Using testHardlink As New DisposableFile(hardlinkPath, False)
                    Return TestNumber("Hardlinks7", WalkmanLib.GetHardlinkCount(testFileSource), 2)
                End Using
            End Using
        End Function

        Function Test_Hardlink8(rootTestFolder As String) As Boolean
            Using testFileSource As New DisposableFile(Path.Combine(rootTestFolder, "hardlinks8Source.txt"))
                Dim hardlinkPath As String = Path.Combine(rootTestFolder, "hardlinks8.txt")

                WalkmanLib.CreateHardLink(hardlinkPath, testFileSource)

                Using testHardlink As New DisposableFile(hardlinkPath, False)
                    Dim lstLinks As List(Of String) = WalkmanLib.GetHardlinkLinks(hardlinkPath).ToList

                    Return TestNumber("Hardlinks8", lstLinks.Count, 2)
                End Using
            End Using
        End Function

        Function Test_Hardlink9(rootTestFolder As String) As Boolean
            Using testFileSource As New DisposableFile(Path.Combine(rootTestFolder, "hardlinks9Source.txt"))
                Dim hardlinkPath As String = Path.Combine(rootTestFolder, "hardlinks9.txt")

                WalkmanLib.CreateHardLink(hardlinkPath, testFileSource)

                Using testHardlink As New DisposableFile(hardlinkPath, False)
                    Dim lstLinks As List(Of String) = WalkmanLib.GetHardlinkLinks(hardlinkPath).ToList
                    Dim linkPath As String = lstLinks.First(Function(x As String)
                                                                Return x.EndsWith("Source.txt")
                                                            End Function)

                    Return TestString("Hardlinks9", linkPath, testFileSource)
                End Using
            End Using
        End Function

        Function Test_Hardlink10(rootTestFolder As String) As Boolean
            Using testFileSource As New DisposableFile(Path.Combine(rootTestFolder, "hardlinks10Source.txt"))
                Dim hardlinkPath As String = Path.Combine(rootTestFolder, "hardlinks10.txt")

                WalkmanLib.CreateHardLink(hardlinkPath, testFileSource)

                Using testHardlink As New DisposableFile(hardlinkPath, False)
                    Dim lstLinks As List(Of String) = WalkmanLib.GetHardlinkLinks(testFileSource).ToList
                    Dim linkPath As String = lstLinks.First(Function(x As String)
                                                                Return x.EndsWith("Source.txt")
                                                            End Function)

                    Return TestString("Hardlinks10", linkPath, testFileSource)
                End Using
            End Using
        End Function

        Function Test_HardlinkThrows5(rootTestFolder As String) As Boolean
            Dim ex As Exception = New NoException
            Try
                WalkmanLib.GetHardlinkCount(Path.Combine(rootTestFolder, "nonExistantFile.txt"))
            Catch ex2 As Exception
                ex = ex2
            End Try
            Return TestType("HardlinkThrows5", ex.GetType, GetType(FileNotFoundException))
        End Function

        Function Test_HardlinkThrows6(rootTestFolder As String) As Boolean
            Dim ex As Exception = New NoException
            Try
                WalkmanLib.GetHardlinkCount(Path.Combine(rootTestFolder, "nonExistantFolder", "nonExistantFile.txt"))
            Catch ex2 As Exception
                ex = ex2
            End Try
            Return TestType("HardlinkThrows6", ex.GetType, GetType(DirectoryNotFoundException))
        End Function

        Function Test_HardlinkThrows7() As Boolean
            Dim ex As Exception = New NoException
            Try
                WalkmanLib.GetHardlinkCount(Path.Combine(Environment.GetEnvironmentVariable("WinDir"), "MEMORY.DMP"))
            Catch ex2 As Exception
                ex = ex2
            End Try
            Return TestType("HardlinkThrows7", ex.GetType, GetType(UnauthorizedAccessException))
        End Function

        Function Test_HardlinkThrows8() As Boolean
            Dim ex As Exception = New NoException
            Try
                WalkmanLib.GetHardlinkCount(Path.Combine(Environment.GetEnvironmentVariable("SystemDrive") & Path.DirectorySeparatorChar, "pagefile.sys"))
            Catch ex2 As Exception
                ex = ex2
            End Try
            Return TestType("HardlinkThrows8", ex.GetType, GetType(IOException))
        End Function

        Function Test_HardlinkThrows9(rootTestFolder As String) As Boolean
            Dim ex As Exception = New NoException
            Try
                WalkmanLib.GetHardlinkLinks(Path.Combine(rootTestFolder, "nonExistantFile.txt")).ToList
            Catch ex2 As Exception
                ex = ex2
            End Try
            Return TestType("HardlinkThrows9", ex.GetType, GetType(FileNotFoundException))
        End Function

        Function Test_HardlinkThrows10(rootTestFolder As String) As Boolean
            Dim ex As Exception = New NoException
            Try
                WalkmanLib.GetHardlinkLinks(Path.Combine(rootTestFolder, "nonExistantFolder", "nonExistantFile.txt")).ToList
            Catch ex2 As Exception
                ex = ex2
            End Try
            Return TestType("HardlinkThrows10", ex.GetType, GetType(DirectoryNotFoundException))
        End Function

        Function Test_HardlinkThrows11() As Boolean
            Dim ex As Exception = New NoException
            Try
                WalkmanLib.GetHardlinkLinks(Path.Combine(Environment.GetEnvironmentVariable("WinDir"), "MEMORY.DMP")).ToList
            Catch ex2 As Exception
                ex = ex2
            End Try
            Return TestType("HardlinkThrows11", ex.GetType, GetType(UnauthorizedAccessException))
        End Function

        Function Test_HardlinkThrows12() As Boolean
            Dim ex As Exception = New NoException
            Try
                WalkmanLib.GetHardlinkLinks(Path.Combine(Environment.GetEnvironmentVariable("SystemDrive") & Path.DirectorySeparatorChar, "pagefile.sys")).ToList
            Catch ex2 As Exception
                ex = ex2
            End Try
            Return TestType("HardlinkThrows12", ex.GetType, GetType(IOException))
        End Function
    End Module
End Namespace

File: Tests/Test_Hardlinks.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace Tests {
    static class Tests_Hardlinks {
        public static bool Test_Hardlinks1(string rootTestFolder) {
            using (var testFileSource = new DisposableFile(Path.Combine(rootTestFolder, "hardlinks1Source.txt"))) {
                File.WriteAllText(testFileSource, "testText");
                string hardlinkPath = Path.Combine(rootTestFolder, "hardlinks1.txt");
                Environment.CurrentDirectory = rootTestFolder;

                WalkmanLib.CreateHardLink(hardlinkPath, "hardlinks1Source.txt");

                using (var testHardlink = new DisposableFile(hardlinkPath, false)) {
                    return GeneralFunctions.TestString("Hardlinks1", File.ReadAllText(hardlinkPath), "testText");
                }
            }
        }

        public static bool Test_Hardlinks2(string rootTestFolder) {
            using (var testFileSource = new DisposableFile(Path.Combine(rootTestFolder, "hardlinks2Source.txt"))) {
                File.WriteAllText(testFileSource, "testText");
                string hardlinkPath = Path.Combine(rootTestFolder, "hardlinks2.txt");

                WalkmanLib.CreateHardLink(hardlinkPath, testFileSource);

                using (var testHardlink = new DisposableFile(hardlinkPath, false)) {
                    return GeneralFunctions.TestString("Hardlinks2", File.ReadAllText(hardlinkPath), "testText");
                }
            }
        }

        public static bool Test_Hardlinks3(string rootTestFolder) {
            using (var testDirRoot = new DisposableDirectory(Path.Combine(rootTestFolder, "hardlinks3Root")))
            using (var testFileSource = new DisposableFile(Path.Combine(testDirRoot, "hardlinks3Source.txt"))) {

                File.WriteAllText(testFileSource, "testText");
                string hardlinkPath = Path.Combine(rootTestFolder, "hardlinks3.txt");
                string hardlinkTarget = Path.Combine(Path.GetFileName(testDirRoot), Path.GetFileName(testFileSource));
                Environment.CurrentDirectory = rootTestFolder;

                WalkmanLib.CreateHardLink(hardlinkPath, hardlinkTarget);

                using (var testHardlink = new DisposableFile(hardlinkPath, false)) {
                    return GeneralFunctions.TestString("Hardlinks3", File.ReadAllText(hardlinkPath), "testText");
                }
            }
        }

        public static bool Test_Hardlinks4(string rootTestFolder) {
            using (var testFileSource = new DisposableFile(Path.Combine(rootTestFolder, "hardlinks4Source.txt")))
            using (var testDirTarget = new DisposableDirectory(Path.Combine(rootTestFolder, "hardlinks4Target"))) {

                File.WriteAllText(testFileSource, "testText");
                string hardlinkPath = Path.Combine(testDirTarget, "hardlinks4.txt");
                string hardlinkTarget = Path.Combine("..", Path.GetFileName(testFileSource));
                Environment.CurrentDirectory = testDirTarget;

                WalkmanLib.CreateHardLink(hardlinkPath, hardlinkTarget);

                Environment.CurrentDirectory = rootTestFolder; // allow testDirTarget to be deleted
                using (var testHardlink = new DisposableFile(hardlinkPath, false)) {
                    return GeneralFunctions.TestString("Hardlinks4", File.ReadAllText(hardlinkPath), "testText");
                }
            }
        }

        public static bool Test_Hardlinks5(string rootTestFolder) {
            using (var testDirSource = new DisposableDirectory(Path.Combine(rootTestFolder, "hardlinks5Source")))
            using (var testFileSource = new DisposableFile(Path.Combine(testDirSource, "hardlinks5Source.txt")))
            using (var testDirTarget = new DisposableDirectory(Path.Combine(rootTestFolder, "hardlinks5Target"))) {

                File.WriteAllText(testFileSource, "testText");
                string hardlinkPath = Path.Combine(testDirTarget, "hardlinks5.txt");
                string hardlinkTarget = Path.Combine("..", Path.GetFileName(testDirSource), Path.GetFileName(testFileSource));
                Environment.CurrentDirectory = testDirTarget;

                WalkmanLib.CreateHardLink(hardlinkPath, hardlinkTarget);

                Environment.CurrentDirectory = rootTestFolder; // allow testDirTarget to be deleted
                using (var testHardlink = new DisposableFile(hardlinkPath, false)) {
                    return GeneralFunctions.TestString("Hardlinks5", File.ReadAllText(hardlinkPath), "testText");
                }
            }
        }

        public static bool Test_HardlinkThrows1(string rootTestFolder) {
            Exception ex = new NoException();
            try {
                WalkmanLib.CreateHardLink(Path.Combine(rootTestFolder, "hardlinkThrows1.txt"), "nonExistantFile.txt");
            } catch (Exception ex2) {
                ex = ex2;
            }
            return GeneralFunctions.TestType("HardlinkThrows1", ex.GetType(), typeof(FileNotFoundException));
        }

        public static bool Test_HardlinkThrows2(string rootTestFolder) {
            using (var testFile = new DisposableFile(Path.Combine(rootTestFolder, "hardlinkThrows2Source.txt"))) {
                Exception ex = new NoException();
                try {
                    WalkmanLib.CreateHardLink(Path.Combine(rootTestFolder, "nonExistantFolder", "hardlinkThrows2.txt"), testFile);
                } catch (Exception ex2) {
                    ex = ex2;
                }
                return GeneralFunctions.TestType("HardlinkThrows2", ex.GetType(), typeof(DirectoryNotFoundException));
            }
        }

        public static bool Test_HardlinkThrows3(string rootTestFolder) {
            using (var testSource = new DisposableFile(Path.Combine(rootTestFolder, "hardlinkThrows3Source.txt")))
            using (var testFile = new DisposableFile(Path.Combine(rootTestFolder, "hardlinkThrows3.txt"))) {
                Exception ex = new NoException();
                try {
                    WalkmanLib.CreateHardLink(testFile, testSource);
                } catch (Exception ex2) {
                    ex = ex2;
                }
                return GeneralFunctions.TestType("HardlinkThrows3", ex.GetType(), typeof(IOException));
            }
        }

        public static bool Test_HardlinkThrows4(string rootTestFolder) {
            using (var testFile = new DisposableFile(Path.Combine(rootTestFolder, "hardlinkThrows4Source.txt"))) {
                Exception ex = new NoException();
                try {
                    WalkmanLib.CreateHardLink(Path.Combine(Environment.SystemDirectory, "symlinkThrows3.txt"), testFile);
                } catch (Exception ex2) {
                    ex = ex2;
                }
                return GeneralFunctions.TestType("HardlinkThrows4", ex.GetType(), typeof(UnauthorizedAccessException));
            }
        }

        public static bool Test_Hardlink6(string rootTestFolder) {
            using (var testFileSource = new DisposableFile(Path.Combine(rootTestFolder, "hardlinks6Source.txt"))) {
                string hardlinkPath = Path.Combine(rootTestFolder, "hardlinks6.txt");

                WalkmanLib.CreateHardLink(hardlinkPath, testFileSource);

                using (var testHardlink = new DisposableFile(hardlinkPath, false)) {
                    return GeneralFunctions.TestNumber("Hardlinks6", WalkmanLib.GetHardlinkCount(hardlinkPath), 2);
                }
            }
        }

        public static bool Test_Hardlink7(string rootTestFolder) {
            using (var testFileSource = new DisposableFile(Path.Combine(rootTestFolder, "hardlinks7Source.txt"))) {
                string hardlinkPath = Path.Combine(rootTestFolder, "hardlinks7.txt");

                WalkmanLib.CreateHardLink(hardlinkPath, testFileSource);

                using (var testHardlink = new DisposableFile(hardlinkPath, false)) {
                    return GeneralFunctions.TestNumber("Hardlinks7", WalkmanLib.GetHardlinkCount(testFileSource), 2);
                }
            }
        }

        public static bool Test_Hardlink8(string rootTestFolder) {
            using (var testFileSource = new DisposableFile(Path.Combine(rootTestFolder, "hardlinks8Source.txt"))) {
                string hardlinkPath = Path.Combine(rootTestFolder, "hardlinks8.txt");

                WalkmanLib.CreateHardLink(hardlinkPath, testFileSource);

                using (var testHardlink = new DisposableFile(hardlinkPath, false)) {
                    List<string> lstLinks = WalkmanLib.GetHardlinkLinks(hardlinkPath).ToList();

                    return GeneralFunctions.TestNumber("Hardlinks8", lstLinks.Count, 2);
                }
            }
        }

        public static bool Test_Hardlink9(string rootTestFolder) {
            using (var testFileSource = new DisposableFile(Path.Combine(rootTestFolder, "hardlinks9Source.txt"))) {
                string hardlinkPath = Path.Combine(rootTestFolder, "hardlinks9.txt");

                WalkmanLib.CreateHardLink(hardlinkPath, testFileSource);

                using (var testHardlink = new DisposableFile(hardlinkPath, false)) {
                    List<string> lstLinks = WalkmanLib.GetHardlinkLinks(hardlinkPath).ToList();
                    string linkPath = lstLinks.First((string x) => x.EndsWith("Source.txt"));

                    return GeneralFunctions.TestString("Hardlinks9", linkPath, testFileSource);
                }
            }
        }

        public static bool Test_Hardlink10(string rootTestFolder) {
            using (var testFileSource = new DisposableFile(Path.Combine(rootTestFolder, "hardlinks10Source.txt"))) {
                string hardlinkPath = Path.Combine(rootTestFolder, "hardlinks10.txt");

                WalkmanLib.CreateHardLink(hardlinkPath, testFileSource);

                using (var testHardlink = new DisposableFile(hardlinkPath, false)) {
                    List<string> lstLinks = WalkmanLib.GetHardlinkLinks(testFileSource).ToList();
                    string linkPath = lstLinks.First((string x) => x.EndsWith("Source.txt"));

                    return GeneralFunctions.TestString("Hardlinks10", linkPath, testFileSource);
                }
            }
        }

        public static bool Test_HardlinkThrows5(string rootTestFolder) {
            Exception ex = new NoException();
            try {
                WalkmanLib.GetHardlinkCount(Path.Combine(rootTestFolder, "nonExistantFile.txt"));
            } catch (Exception ex2) {
                ex = ex2;
            }
            return GeneralFunctions.TestType("HardlinkThrows5", ex.GetType(), typeof(FileNotFoundException));
        }

        public static bool Test_HardlinkThrows6(string rootTestFolder) {
            Exception ex = new NoException();
            try {
                WalkmanLib.GetHardlinkCount(Path.Combine(rootTestFolder, "nonExistantFolder", "nonExistantFile.txt"));
            } catch (Exception ex2) {
                ex = ex2;
            }
            return GeneralFunctions.TestType("HardlinkThrows6", ex.GetType(), typeof(DirectoryNotFoundException));
        }

        public static bool Test_HardlinkThrows7() {
            Exception ex = new NoException();
            try {
                WalkmanLib.GetHardlinkCount(Path.Combine(Environment.GetEnvironmentVariable("WinDir"), "Temp", "MpCmdRun.log"));
            } catch (Exception ex2) {
                ex = ex2;
            }
            return GeneralFunctions.TestType("HardlinkThrows7", ex.GetType(), typeof(UnauthorizedAccessException));
        }

        public static bool Test_HardlinkThrows8() {
            Exception ex = new NoException();
            try {
                WalkmanLib.GetHardlinkCount(Path.Combine(Environment.GetEnvironmentVariable("SystemDrive") + Path.DirectorySeparatorChar, "pagefile.sys"));
            } catch (Exception ex2) {
                ex = ex2;
            }
            return GeneralFunctions.TestType("HardlinkThrows8", ex.GetType(), typeof(IOException));
        }

        public static bool Test_HardlinkThrows9(string rootTestFolder) {
            Exception ex = new NoException();
            try {
                WalkmanLib.GetHardlinkLinks(Path.Combine(rootTestFolder, "nonExistantFile.txt")).ToList();
            } catch (Exception ex2) {
                ex = ex2;
            }
            return GeneralFunctions.TestType("HardlinkThrows9", ex.GetType(), typeof(FileNotFoundException));
        }

        public static bool Test_HardlinkThrows10(string rootTestFolder) {
            Exception ex = new NoException();
            try {
                WalkmanLib.GetHardlinkLinks(Path.Combine(rootTestFolder, "nonExistantFolder", "nonExistantFile.txt")).ToList();
            } catch (Exception ex2) {
                ex = ex2;
            }
            return GeneralFunctions.TestType("HardlinkThrows10", ex.GetType(), typeof(DirectoryNotFoundException));
        }

        public static bool Test_HardlinkThrows11() {
            Exception ex = new NoException();
            try {
                WalkmanLib.GetHardlinkLinks(Path.Combine(Environment.GetEnvironmentVariable("WinDir"), "Temp", "MpCmdRun.log")).ToList();
            } catch (Exception ex2) {
                ex = ex2;
            }
            return GeneralFunctions.TestType("HardlinkThrows11", ex.GetType(), typeof(UnauthorizedAccessException));
        }

        public static bool Test_HardlinkThrows12() {
            Exception ex = new NoException();
            try {
                WalkmanLib.GetHardlinkLinks(Path.Combine(Environment.GetEnvironmentVariable("SystemDrive") + Path.DirectorySeparatorChar, "pagefile.sys")).ToList();
            } catch (Exception ex2) {
                ex = ex2;
            }
            return GeneralFunctions.TestType("HardlinkThrows12", ex.GetType(), typeof(IOException));
        }
    }
}
